src/python/pants/backend/codegen/tasks/simple_codegen_task.py

```diff
@@ -172,3 +217,119 @@ def execute(self):
 
       if self.artifact_cache_writes_enabled():
         self.update_artifact_cache(vts_artifactfiles_pairs)
+
+
+  class CodegenStrategy(object):
+    """Abstract strategies for running codegen.
+
+    Includes predicting generated sources, partitioning targets for execution, etc.
+    """
+    __metaclass__ = ABCMeta
+
+    @property
+    def name(self):
+      """The name of this strategy (eg, 'isolated')."""
+      raise NotImplementedError
+
+    @abstractmethod
+    def execute_codegen(self, targets):
+      """Invokes the task's execute_codegen on the targets.
+
+      Subclasses decide how the targets are partitioned before being sent to the task's
+      execute_codegen method.
+      :targets: a set of targets.
+      """
+      raise NotImplementedError
```

Comment:
No need to `raise NotImplementedError` in an `@abstractmethod` annotated function